tensorrt_llm/serve/openai_disagg_service.py

```diff
@@ -0,0 +1,329 @@
+                )
+            await self._wait_for_all_servers_ready()
+
+    async def teardown(self) -> None:
+        await self._ctx_client.shutdown()
+        await self._gen_client.shutdown()
+
+        if self._disagg_cluster_manager:
+            await self._disagg_cluster_manager.stop()
+
+        if self._metadata_server:
+            await self._ctx_router.stop_server_monitoring()
+            await self._gen_router.stop_server_monitoring()
+
+    async def _wait_for_all_servers_ready(self) -> None:
+        async def check_servers_ready():
+            elapsed_time = 0
+            interval = 3
+            while elapsed_time < self._server_start_timeout_secs:
+                _, unready_ctx_servers = await self._ctx_client.check_ready()
+                _, unready_gen_servers = await self._gen_client.check_ready()
+                if len(unready_ctx_servers) == 0 and len(unready_gen_servers) == 0:
+                    logger.info("All servers are ready")
+                    return
+                logger.info(
+                    f"Waiting for servers, context: {unready_ctx_servers}, generation: {unready_gen_servers}"
+                )
+                await asyncio.sleep(interval)
+                elapsed_time += interval
+
+        try:
+            await asyncio.wait_for(check_servers_ready(), timeout=self._server_start_timeout_secs)
+        except asyncio.TimeoutError:
+            raise TimeoutError("Timeout waiting for context and generation servers to be ready")
+
+    async def _on_worker_event(self, worker_info: WorkerInfo, event_type: WatchEventType):
+        router_map = {ServerRole.CONTEXT: self._ctx_router, ServerRole.GENERATION: self._gen_router}
+        worker_addr = f"{worker_info.host}:{worker_info.port}"
+        router = router_map[worker_info.role]
+        if event_type == WatchEventType.SET:
+            await router.add_server(worker_addr)
+        elif event_type == WatchEventType.DELETE:
+            await router.remove_server(worker_addr)
+        logger.info(f"Worker {event_type.name} event: {worker_info.worker_id}, {worker_addr}")
+
+    async def _verify_ctx_response(self, ctx_response: UCompletionResponse) -> None:
+        if ctx_response:
+            if len(ctx_response.choices) != 1:
+                raise ValueError(
+                    f"Context server returned {len(ctx_response.choices)} choices, expecting 1."
+                )
+            if ctx_response.choices[0].disaggregated_params is None:
+                raise ValueError("Context server did not return disaggregated params")
+            if ctx_response.choices[0].disaggregated_params.ctx_request_id is None:
+                raise ValueError("Invalid disaggregated params in context phase response.")
+            return ctx_response
+
+
+# FIXME: This is a demo to show the basic idea of disagg-service with pre-allocating generation
+class OpenAIDisaggregatedPreAllocService(OpenAIDisaggregatedService):
```

Comment:
I think it's great we have this prototype but I would move to a separate MR with tests to verify it's working as expected.